controllers/controllers.go

```diff
@@ -66,3 +79,39 @@ func mergeMap(target map[string]string, desired map[string]string) bool {
 	}
 	return changed
 }
+
+// Requeue provides a wrapper around different results from a subreconciler.
+type Requeue struct {
+	// Delay provides an optional delay before requeueing reconciliattion.
+	Delay time.Duration
+
+	// Error provides an error that we encountered that forced a requeue.
+	Error error
+
+	// Message provides a log message that explains the reason for the requeue.
+	Message string
+}
+
+// processRequeue interprets a requeue resulit from a subreconciler.
+func processRequeue(requeue *Requeue, subReconciler interface{}, object runtime.Object, recorder record.EventRecorder, logger logr.Logger) (ctrl.Result, error) {
+	if requeue.Message == "" && requeue.Error != nil {
+		requeue.Message = requeue.Error.Error()
+	}
+
+	err := requeue.Error
+	if err != nil && k8serrors.IsConflict(err) {
+		err = nil
+		if requeue.Delay == time.Duration(0) {
+			requeue.Delay = time.Minute
+		}
+	}
+
+	logger.Info("Reconciliation terminated early", "subReconciler", fmt.Sprintf("%T", subReconciler), "message", requeue.Message, "requeueAfter", requeue.Delay)
+	recorder.Event(object, corev1.EventTypeNormal, "ReconciliationEndedEarly", requeue.Message)
+
+	if err != nil {
+		logger.Error(err, "Error in reconciliation", "subReconciler", fmt.Sprintf("%T", subReconciler))
```

Comment:
With that change we will print the error twice once in the Info statement above and a second time here. Assuming that most requeue structs won't set the message when an error occurs.